fhirstore/fhirstore.py

```diff
@@ -76,55 +98,50 @@ def bootstrap(self, depth=3, resource=None, show_progress=True):
-            collections = tqdm(
-                collections, file=sys.stdout, desc="Loading collections from database...",
-            )
-
-        for collection in collections:
-            json_schema = self.db.get_collection(collection).options()["validator"]["$jsonSchema"]
-            self.resources[collection] = json_schema
 
     def validate_resource_type(self, resource_type):
-        if resource_type is None:
+        if not resource_type:
             raise BadRequestError("resourceType is missing in resource")
-
         elif resource_type not in self.resources:
             raise NotFoundError(f'unsupported FHIR resource: "{resource_type}"')
 
-    def create(self, resource, bypass_document_validation=False):
+    def normalize_resource(self, resource: Union[Dict, FHIRAbstractModel]) -> FHIRAbstractModel:
+        if isinstance(resource, dict):
+            resource_type = resource.get("resourceType")
+            self.validate_resource_type(resource_type)
+            resource = construct_fhir_element(resource_type, resource)
+
+        return resource
+
+    def error(self, errors: List[str], severity="error", code="invalid") -> OperationOutcome:
+        issues = [{"severity": severity, "code": code, "diagnostics": err} for err in errors]
+        return OperationOutcome(**{"issue": issues})
+
+    def create(self, resource: Union[Dict, FHIRAbstractModel], bypass_document_validation=False):
         """
         Creates a resource. The structure of the resource will be checked
         against its json-schema FHIR definition.
 
         Args:
-            - resource_type: type of the resource (eg: 'Patient')
-            - id: The expected id is the resource 'id', not the
-                  internal database identifier ('_id').
+            - resource: dict with the resource data
 
-        Returns: The updated resource.
+        Returns: The created resource.
         """
-        resource_type = resource.get("resourceType")
-        self.validate_resource_type(resource_type)
-
         try:
-            res = self.db[resource_type].insert_one(
-                resource, bypass_document_validation=bypass_document_validation
+            resource = self.normalize_resource(resource)
+        except pydantic.ValidationError as e:
+            return self.error(
+                [
+                    f"{err['msg'] or 'Validation error'}: "
+                    f"{e.model.get_resource_type()}.{'.'.join([str(l) for l in err['loc']])}"
+                    for err in e.errors()
+                ]
             )
-            return {**resource, "_id": res.inserted_id}
-        except DuplicateKeyError as e:
-            raise e
-        except WriteError:
-            self.validate(resource)
+
+        res = self.db[resource.resource_type].insert_one(
+            json.loads(resource.json()), bypass_document_validation=bypass_document_validation
+        )
+        return {**resource.dict(), "_id": res.inserted_id}
```

Comment:
What if resource is a FHIRAbstractModel?

Reply:
it's always a FHIRAbstractModel at this point, am I missing something ?

Reply:
My bad, self.normalize_resource doesn't do anything if resource is already a FHIRAbstractModel 👍
Maybe add a comment about that and/or docstring on self.normalize_resource. It's ok if we don't do it right now though